k8s: Migrate policy watchers to Cell + Resource

gandro opened the pull request:
This PR moves the policy related K8s resources, i.e. Kubernetes `NetworkPolicy` (KNP), `CiliumNetworkPolicy` (CNP), `ClusterwideCiliumNetworkPolicy` (CCNP) and `CiliumCIDRGroup` (CCG) from `pkg/k8s/watchers` into it's own cell in `pkg/policy/k8s`.

I've been careful to check that this should not have any externally visible changes. Importantly in that regard, `K8sWatcher.InitK8sSubsystem` still waits for all policy related resources to be synced by the new cell before continuing daemon startup.

During the refactor, I noticed a lot of dead/unused code in the affected areas. I have removed all such code in separate commits. This makes the diff look a bit scary, but I've attempted to explain the the commit message why the removal is safe. Because of this, please **review per commit**.

There is more to be done (e.g. reworking `k8s.PreprocessRules`), but because the PR is already big enough as is, I'll defer further cleanup to future PRs.

Ref: #30162

gandro committed:
k8s: Provide `synced` types via Cell

This commit introduces a new Cell which provides the `synced.Resources`
and `synced.APIGroup` types via a Cell. This will allow other cells to
access those resources in conjunction with the existing K8s watcher.

This commit by itself should not contain any functional changes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/watcher: Move CNP/CCNP/CCG watcher into struct

This is a preparatory commit to move all the state associated with the
Cilium Network Policy resources into a separate struct. This allows the
logic to be moved into a separate package in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s: Move KNP/CNP/CCNP/CCG watchers into cell

This commit moves the network policy related K8s resource watchers out
of `pkg/k8s/watchers` and into a separate cell. Because daemon startup
still requires policy resources to be synced before continuing, we still
wait for the policy resources in `InitK8sSubsystem`. By sharing the
`synced.Resources` type between the legacy K8s watcher and
`policy/k8s.Cell`, the sync event is still respected as before.

All other code is moved more or less unmodified to preserve its git
history.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/hubble: Remove GetStore helper

This commit removes the `GetStore` interface from both Hubble and the
K8s watcher. The `GetStore` method has not been used in tree for years
and is no longer required, as out-of-tree code which requires access to
K8s stores can now `Resource[T].Store()` instead.

The network policy store is completely removed with this commit, where
as the pod store infrastucture remains accessible via `GetCachedPod`,
which still has users in-tree.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/watchers: Unify KNP with CNP logic

This commit changes the initialization of the KNP watcher to mirror more
closely how the CNP watcher works, i.e. making the `init` function
private and using a context as a stop signal.

Using a context here instead of a stop channel is fine because the
context is stopped when the Daemon is stopped.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/watcher: Move KNP watcher into struct

This is a preparatory commit to also move the Kubernetes Network Policy
resource into the PolicyWatcher struct. This allows the logic to be
moved into a separate package in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/watcher: Removed unused constants

These constants are no longer in use.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Remove unused parameters

This removes unused metrics label and clientset parameters from private
functions in the Kubernetes policy watcher package.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s: Remove unused CNP controller name

Commit 2835b063800f ("Delete deprecated CNPStatusUpdates and
K8sEventHandover") removed the creation of the CNP status update
controller. Since there is now no longer any controller started for a
CNP, we can safely remove the code which attempts to stop the
(non-existant) controller. In addition, we can also remove the
controller name helper.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Remove unused policy import tracking

This commit removes the `importMetadataCache` infrastructure, as it is
no longer used. The two remaining calls were a upsert() followed by a
delete(), but those did not have any visible side-effects. Thus it is
safe to remove them, as the tracking is clearly not used anymore.

The last real use of `importMetadataCache` was removed a few weeks ago
in commit 2835b063800f ("Delete deprecated CNPStatusUpdates and
K8sEventHandover"), which also removed the `importMetadataCache.get()`
helper.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

bimmlerd commented on pkg/policy/k8s/cell.go:
```diff
@@ -0,0 +1,107 @@
+	PolicyDelete(labels labels.LabelArray, opts *policy.DeleteOptions) (newRev uint64, err error)
+}
+
+type WatcherParams struct {
```
For consistency

```suggestion
type PolicyWatcherParams struct {
```

bimmlerd commented on pkg/policy/k8s/cell.go:
```diff
@@ -0,0 +1,107 @@
+	k8sAPIGroupCiliumCIDRGroupV2Alpha1          = "cilium/v2alpha1::CiliumCIDRGroup"
+)
+
+var Cell = cell.Module(
```
doc comment here would be nice - it's where code navigation dumps you when looking up this cell

bimmlerd commented on pkg/policy/k8s/cell.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "k8s-policy-watcher")
```
I'm understanding you left the log as is to not make this PR bigger?

Note that the logging is calling it `k8s-policy-watcher` whereas the module is called `policy-k8s-watcher`. If you intend to transform the logging at some point, that module name will become the subsystem name, which would hence change. Are we fine with that?

gandro replied:
> I'm understanding you left the log as is to not make this PR bigger?

Yes, though I think I might as well transform that too if the diff is not too big (since follow-up PRs are likely going to focus on different areas of the code).

gandro replied:
Pushed an additional commit which uses the hive based logging.

bimmlerd commented on daemon/cmd/cells.go:
```diff
@@ -208,6 +215,15 @@ var (
 
+		// K8s policy resource watcher cell. It depends on the daemon which we cast into
+		// the interface type here to avoid a circular package import
+		cell.Provide(func(p promise.Promise[*Daemon]) promise.Promise[policyK8s.PolicyManager] {
```
Do you also have a plan in mind for how to factor out the PolicyManager things out of the daemon, so that this isn't needed anymore?

gandro replied:
It is on our radar (c.f. https://github.com/cilium/cilium/issues/30162) but it's a lower priority for us at the moment.

bimmlerd reviewed (approved):
very nice! couple of minor things, but nothing blocking

gandro committed:
policy/k8s: Switch to hive-provided logger

This is the idiomatic way on how to log in a hive cell.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

squeed commented on pkg/policy/k8s/cilium_network_policy.go:
```diff
@@ -84,7 +55,7 @@ func (k *K8sWatcher) ciliumNetworkPoliciesInit(ctx context.Context, cs client.Cl
 		cnpCache := make(map[resource.Key]*types.SlimCNP)
```
Let's lift these caches to a field on the PolicyWatcher. That should simplify method signatures nicely.

squeed commented on pkg/policy/k8s/cilium_network_policy.go:
```diff
@@ -84,7 +55,7 @@ func (k *K8sWatcher) ciliumNetworkPoliciesInit(ctx context.Context, cs client.Cl
+		cidrGroupEvents := p.CiliumCIDRGroups.Events(ctx)
 
 		// cidrGroupPolicies is the set of policies that are referencing CiliumCIDRGroup objects.
 		cidrGroupPolicies := make(map[resource.Key]struct{})
```
Likewise, let's move this to a struct field too.

gandro committed:
policy/k8s: Move shared caches into PolicyWatcher struct

This commit moves the CNP and CCG caches into the PolicyWatcher struct.
This allows us to simplify various function and method signatures, as
the caches no longer need to be passed in from the outside.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Move PolicyWatcher into separate file

The policy watcher is shared for all watched resources. Therefore move
it out of the CNP file into a separate file. This commit contians no
changes to the function body.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Unify KNP and CNP handling

This merges the KNP resource watcher go routine with the CNP one. This
ensures that no `PolicyWatcher` fields are accessed concurrently and
unifies the code a bit:

1. The "event timestamp" is now set after the operation has been
   completed, which is how the CNP methods already did it.
2. The no-op stoppable wait group is removed and the API group is
   registered via `registerResourceWithSyncFn` helper.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

kaworu reviewed (changes requested):
Hubble changes LGTM, but I think [some bits were missed in `local_observer_test.go`](https://github.com/cilium/cilium/blob/661393587e164eb6184135c2ebfe1bdc9bbbc86d/pkg/hubble/observer/local_observer_test.go#L497-L499)

squeed reviewed (approved):
super nice, dig it.

christarazi reviewed (approved):
:rocket:

nathanjsweet reviewed (approved):
Nice!
